### ISSUE TITLE
event condition never satisfied

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#886]
I started using an ItemStateEvent Condition, but it is never satisfied.
After some debugging I have a look at the "isSatisfied" method of the EventConditionHandler.

The code looks like:
```java
        Event event = inputs.get("event") != null ? (Event) inputs.get("event") : null;
        if (event != null) {
            return isConfiguredAndMatches(TOPIC, event.getTopic()) && isConfiguredAndMatches(SOURCE, event.getSource())
                    && isConfiguredAndMatches(PAYLOAD, event.getPayload())
                    && isConfiguredAndMatches(EVENTTYPE, event.getType());
        }
        return false;
```

If I looked at the content of <code>inputs</code> my event is not named "event", but "trigger_1.event". If I changed the strings, the condition is working.
I assume the naming "event" is correct and the function that fills the map have to change the "trigger_X.event" naming to "event".

But I do not know any of the automation code...
Which code has to be fixed?

**Comment by maggu2810** [comment]
```java
Map<String, ?> compositeContext = getCompositeContext(internalContext);
```

The internal and the composite context also using the "trigger_1.event" key.
Should the "getCompositeContext" function remove the prefix of event?

internalContext = {trigger_1.event=sw1 updated to OFF}
compositeContext = {$state=OFF, trigger_1.event=sw1 updated to OFF, $itemName=sw1}

**Comment by maggu2810** [comment]
Not really in context to this issue, but for discussion:

RuleEngine:

```java
    /**
     * @return copy of current context in rule engine
     */
    private Map<String, Object> getContext(String ruleUID) {
        return getContext(ruleUID, null);
    }
```

I don't see at which place the "copy" ("of current context") is done.
We take the context from the contextMap and return a reference to it.

**Comment by maggu2810** [comment]
This is the place the module UID ("trigger_1" in this case) is added:

```java
    private void updateContext(String ruleUID, String moduleUID, Map<String, ?> outputs) {
        Map<String, Object> context = getContext(ruleUID);
        for (Map.Entry<String, ?> entry : outputs.entrySet()) {
            final String key = moduleUID + OUTPUT_SEPARATOR + entry.getKey();
            context.put(key, entry.getValue());
        }
    }
```

**Comment by marinmitev** [comment]
I guess that the condition defined in your rule is lacking the input object

See how it is used in the test module:
		"inputs":{
			"event":"ItemStateChangeTriggerID.event"
		}
at line 32 in file: 
bundles\automation\org.eclipse.smarthome.automation.integration.test\ESH-INF\automation\rules\ItemRuleDefinition.json

If you have declared "event" input in your rule condition it will be put in the rule context - this is the code at line 1048 in RuleEngine.java
```java
private boolean calculateConditions(Rule rule) {
...
            Map<String, Object> context = getContext(rule.getUID(), c.getConnections());
            if (!tHandler.isSatisfied(context)) {
..
```
Additionaly as you have noticed the rule context is filled with all the outputs from other rule modules to cover the scenarious with Script Action where you cannot define the inputs staticaly.
Such an example can be seen in the file \bundles\automation\org.eclipse.smarthome.automation.module.script.test\ESH-INF\automation\rules\DemoScriptRule.json
The script action doesnot have inputs and uses the trigger's event by refering to the id of the trigger
print(trigger.event)

**Comment by maggu2810** [comment]
> I guess that the condition defined in your rule is lacking the input object

All the stuff I am using is the one available in the repository (master branch).
The ItemStateEvent is defined in org.eclipse.smarthome.automation.module.core/ESH-INF/automation/moduletypes/ItemEventConditionModuleTypeDefinition.json

The context keys are prefixed by the module UID (as stated above).

The isSatisfied of the event condition is using a key without the UID (it does not depend on a certain module UID).

I think I am missing something (surely).

Perhaps you could try it yourself, using the current master and an "ItemStateEvent Condition".
Is it working for you?

**Comment by maggu2810** [comment]
> All the stuff I am using is the one available in the repository (master branch).

... and I am using the Paper UI to create a rule.

**Comment by maggu2810** [comment]
The Paper UI shows the rule this way:

```text
{
  "enabled": true,
  "status": {
    "status": "IDLE",
    "statusDetail": "NONE"
  },
  "triggers": [
    {
      "id": "trigger_1",
      "label": "Item State Trigger",
      "description": "This triggers a rule if an items state changed",
      "configuration": {
        "itemName": "sw1"
      },
      "type": "ItemStateChangeTrigger"
    }
  ],
  "conditions": [
    {
      "inputs": {},
      "id": "condition_2",
      "label": "ItemStateEvent Condition",
      "description": "checks if an item event was a state event and compares the new state",
      "configuration": {
        "itemName": "sw1",
        "state": "OFF"
      },
      "type": "ItemStateEventCondition"
    }
  ],
  "actions": [
    {
      "inputs": {},
      "id": "action_3",
      "label": "Post command to an item",
      "description": "posts commands on items",
      "configuration": {
        "itemName": "sw1",
        "command": "ON"
      },
      "type": "ItemPostCommandAction"
    }
  ],
  "configuration": {},
  "configDescriptions": [],
  "uid": "rule_1",
  "name": "Test",
  "tags": []
}
```

**Comment by marinmitev** [comment]
Sorry if my explanation was not clear, but in your json structure for rule_1 you have empty map for the condition inputs:
```
  "conditions": [
    {
      "inputs": {},
```
instead of the empty map, it should be
```
"inputs":{
			"event":"trigger_1.event"
		}
```

**Comment by maggu2810** [comment]
> Sorry if my explanation was not clear

I understand it now.
You have point to the json structure of the generated rule...
Thank you!

I see that the generated rule is missing the correct condition input.

Could you point me to the location the rule generation is done?
Otherwise I will try to find it myself, but I believe you are more familiar with all that code. :wink:

**Comment by marinmitev** [comment]
well, I havent written the code for PaperUI but I guess this is the source code
\extensions\ui\org.eclipse.smarthome.ui.paper\web\js\controllers.rules.js

**Comment by marinmitev** [comment]
Hi @aounhaider1 could you please check this issue with rule generation

**Comment by kaikreuzer** [comment]
You should note that the current rule UI is an initial version that is being worked on.
Mapping input/outputs of modules is not yet supported. Feel free to make suggestions on how this should best be introduced in the UI.
Besides letting the user do the mapping, we also discussed in the past an "auto-mapping" of inputs and outputs. If there is exactly one output that matches the type (and tags) of an input, there should be no need for the user to take any action. I am now wondering if this auto-mapping is a logic that the UI should do and send the final rule to the runtime or if this is logic that the runtime should hold, so that it excepts "unmapped" modules and does the mapping itself if possible (and rejecting the rule otherwise). What do you think?

**Comment by marinmitev** [comment]
automatic mapping is good idea but it is better to be done in the UI - which can ask the user for confirmation before sending the rule to the engine. 
If it is implemented on the server side its behavior will be hidden and error-prone.

**Comment by maggu2810** [comment]
> You should note that the current rule UI is an initial version that is being worked on.

That is fine, but I would prefer to keep non-working stuff tracked here.
For me, I have seen that the Paper UI can create rules and wonder, why the rules are not working.
I did not found an issue (or PR) here, so I am informed, that mapping of I/Os does not work.

> Mapping input/outputs of modules is not yet supported. Feel free to make suggestions on how this should best be introduced in the UI.

I will make suggestions if I have some ideas (as always).

> Besides letting the user do the mapping, we also discussed in the past an "auto-mapping" of inputs and outputs. If there is exactly one output that matches the type (and tags) of an input, there should be no need for the user to take any action.

Sounds good.

> I am now wondering if this auto-mapping is a logic that the UI should do and send the final rule to the runtime or if this is logic that the runtime should hold, so that it excepts "unmapped" modules and does the mapping itself if possible (and rejecting the rule otherwise). What do you think?

My first idea has been to do it in the Java code.
The comment of @marinmitev is okay, but perhaps I could be implemented in a way that make it more transparent. The REST API could provide a function to do that auto-mapping that receive all informations. It could return an error if no auto-mapping could be done.
If no auto-mapping is necessary or it could be done correctly a valid JSON structure could be returned.

The very simple UI could then on save / update call the auto-mapping function (if it needs to be done or not). If that auto-mapping succeeded (again regardless if something has been changed or not) the JSON rule could be saved using the current REST API. If that auto-mapping failed an error page / dialog could be shown.

**Comment by danchom** [comment]
I don't understand auto-mapping idea. Is it some kind of many to many relation between inputs/outputs of modules? Yes we have some discussion about tags and and their meaning for GUI to propose connections, but who will define these tags? If they are defined by the rule engine they will be not suitable for all cases. If they are defined by the users, isn't be easier for them to define correct module links. 
If this auto-mapping is done to create a rule only by two clicks I think it is not serious. If you don't know module types which are you using and can't connect them what rule are you going to do?

**Comment by maggu2810** [comment]
Perhaps auto-mapping is not needed at all. As I started today with all that rule stuff, I have to read much more code and documentation before I can say anything to it.

So, to keep this issue on their topic:
There is a missing feature in the Paper UI that does not map input / outputs.

**Comment by maggu2810** [comment]
I changed the JSON (added the input(s) to the condition) using the REST API and the rule is working.
Thank you @marinmitev for your help.

**Comment by kaikreuzer** [comment]
> automatic mapping is good idea but it is better to be done in the UI

Which means that all potential UIs have to implement complex logic and potentially do it in different ways. Like @maggu2810, I would therefore prefer to have this implemented on the server side.

> I don't understand auto-mapping idea.

We had long discussions about that a year ago, which I won't repeat in detail. But to sum up the idea:
In most cases, the inputs/outputs should actually be hidden and not visible to the user. Taking IFTTT as a guideline: "If I upload a new photo to dropbox, publish it on Flickr". The output of the trigger is a photo, the input of the action is a photo. There is absolutely no need to bother the user with anything...

> There is a missing feature in the Paper UI that does not map input / outputs.

No, there is a missing feature in the rule runtime that it does not map input / outputs :-P

**Comment by danchom** [comment]
Ok, I understand the idea to support functionality of IFTTT, but there are problems to solve when auto-mapping is done. 
The rule engine connects module's input to output of the other module. 
If a rule has two triggers with same tag i.e. temperature (or one trigger which has two outputs both tagged as temperature - one for C and the other for K) and the action has a temperature input. The rule engine has to decide how to connect action input and outputs of the triggers. Other problem is when an input and an output have same tags but the rule must not connect them.
May be we can add auto-mapping property of the rule with default value false. When it is switched on the rule engine will try to link input and outputs with same tags, but in case of conflict the rule validation will fail.

**Comment by maggu2810** [comment]
> No, there is a missing feature in the rule runtime that it does not map input / outputs :-P

 You won :wink: 

> May be we can add auto-mapping property of the rule with default value false.

Why not using a separate call as mentioned above?
So the UI could do more stuff e.g. showing the user what will be changed by the auto-mapping and if he want to use that one, etc.

**Comment by kaikreuzer** [comment]
> Why not using a separate call as mentioned above?

+1. It could map what is possible to map and leave the rest unmapped.
Another feature on the REST API could be a filtering of module types, whether they are auto-mappable on a given set of outputs. This would allow the UI to only offer the modules to the user that make sense for a certain rule.

**Comment by danchom** [comment]
Sorry but I don't understand your proposal for auto-mapping. To have additional rest add call which do it? May be it is better to add a new add method in RuleRegitry. In that way not only the rest call  but other modules like console commands and java api will be able to use this functionality.

**Comment by maggu2810** [comment]
Sure, it should not be implemented in the REST interface but the function should be provided by the REST interface. I don't think that someone here wants to implement this one for REST only.

**Comment by danchom** [comment]
Ok, It is time for auto-mapping. May be this issue has to be closed and open a new one about auto-mapping. We have to decide:
1) It has to be implemented in RuleEngine - because rest, java api and console commands can use it.
1.1) a new RuleEngine.addRuleAutoMap(rule) method
1.2) a new parameter to the RuleEngine.addRule(Rule rule, boolean isEnabled, boolean autoMapped)
1.3) add an configuration property (auto-map:true) of added rule
2) The auto-mapping can be done in following way:
2.1) how to interpret tags & or || between them: 
2.1.1) logical and: input is tagged with: (temperature, K) which means it can be connect to outputs  tagged  as (temperature, K, digit), but not to outputs tagged as (temperature), because the output can produce C temperature. This means the input tags must be subset of the output tags.
2.1.2)  logical or:  input is tagged with: (temperature, digit) which means it accepted temperature or any digit value and can be connected to every output tagged as (digit) or tagged as (temperature). In this case the input tags must be super set of the output once to be connected
3) when connections can't be done the rule engine will return rule without connections silently.
4) when a conflict appears (the input can be connected to more then one output), the rule engine has to choose the most specific one (according of 2.1.1 or 2.1.2). In case if they are equal specific the rule engine will select the first one.

**Comment by kaikreuzer** [comment]
1) Are there any situations where we do NOT want auto-mapping? Anyhow, mappings will only be done if there is exactly one clear mapping possible.
2) I think 2.1.1 is what we had agreed upon in the past. Input tags are regarded as mandatory "pre-requisites"
3) ok
4) I would say that no auto-mapping should be done in such a case, see 1)

**Comment by danchom** [comment]
1)It  is ok for men, in case when there is no conflict
2) ok
4) in case of conflict, connection will not be created.

**Comment by danchom** [comment]
May be a problem can appear when connections fo the rule are set when it is created. Then the auto-maping can add additional connections. I think in this case the automaping must not be permitted.

**Comment by kaikreuzer** [comment]
This raises the question if the auto-mapping should be re-done on any rule update as well? Should it also automatically discard invalid connections? Do we have to differentiate between connections that were auto-added and ones that were declared by the user/UI?

**Comment by danchom** [comment]
For me:
>This raises the question if the auto-mapping should be re-done on any rule update as well?
It has to be done only when the rule is created
> Should it also automatically discard invalid connections? 
No because the user decision always has priority. 
>Do we have to differentiate between connections that were auto-added and ones that were declared by the user/UI?
Yes because the user decision always has priority.

**Comment by danchom** [comment]
On the other hand, applying of the auto mapping will not be a problem, if we have a parameter or configuration property or entire new add method.

**Comment by kaikreuzer** [comment]
> It has to be done only when the rule is created

I don't agree - there is no difference between an "initially created rule" and a "completely edited rule that only has the id in common with the previous version" - so for me, it should be applied on any update as well.

> No because the user decision always has priority.

Which means that the REST request would result in an error, because we would not accept invalid mappings, right? This can be awkward, because the user might not have created the connection himself (it was the auto-mapping upon creation), but only edited a detail that made the connection become invalid. So we shift the responsibility to the UI to make sure that invalid connections are automatically removed before submission.

> Yes because the user decision always has priority.

Well, the user automatically has priority, as no auto-mapping will take place if he already created connections himself. The question is rather if there is a use case where the user wants to specify that a certain (obvious and thus auto-mapped) connection MUST NOT be created. I don't really see a convincing use case for that.

**Comment by adimova** [comment]
I see the things like this:
If the user has left some inputs not connected to outputs, then the auto-mapping should be applied on them, if it is possible, no matter if the rule is newly created or updated.
A validation has to be performed in both cases and the rule should be left NOT_INITIALIZED if it fails. Then it will be the user's responsibility to make the rule "valid".
I do not see a reason to differentiate between connections that were auto-added and ones that were declared by the user/UI.

**Comment by maggu2810** [comment]
Hm, what about my commets above?

The REST API could provide a function to do that auto-mapping that receive all informations (the mapping with not be done in the REST implementation, but forwared to). It could return an error if no auto-mapping could be done.
If no auto-mapping is necessary or it could be done correctly a valid JSON structure could be returned.

The very simple UI could then on save / update call the auto-mapping function (if it needs to be done or not). If that auto-mapping succeeded (again regardless if something has been changed or not) the JSON rule could be saved using the current REST API. If that auto-mapping failed an error page / dialog could be shown.

**Comment by danchom** [comment]
My point of view is the auto mapping should be some kind of initial initialization of the rule, in case when the user can't, don't know or is lazy to do it itself. If the auto mapping overwrite user decisions we  can have following problems:
1) the user will never remove connection which is done by auto-mapping
2) the user can receive unexpected results base on created rule. I think it is not unusual to have a module type with defined inputs  which are not connected anywhere. For example we can have a module type of add action with defined 8 inputs (because we want this action to be universal and adds up to 8 (for example) digits). Usually only part of these inputa will be used, the other should be empty. If the auto mapping decides to links these not connected inputs to available inputs , the result of add operation will be unexpected. And the worst  is that the user can't fix the unwanted links because they will be applied again after the update. 
@maggu2810  we are discussing  the case when we have only one add(rule) method and it should do auto mapping by default.

**Comment by adimova** [comment]
> the auto-mapping should be applied on them, if it is possible
I mean that if the inputs are defined as required, has no default value and user has left them unconnected, then the auto-mapping should be applied on them

**Comment by kaikreuzer** [comment]
> The REST API could provide a function to do that auto-mapping that receive all informations 

This means that the UI has always to do at least two calls and that it must not forget the first one. This is imho an option, if we say that auto-mapping is an optional feature that a UI is free to decide to use it or not. If it is mandatory, I would actually not expose it through REST.

> we can have a module type of add action with defined 8 inputs (because we want this action to be universal and adds up to 8 (for example) digits)

The proper design would probably to have a single input with multiple=true.

In general, when talking about "the user" here, we should have in mind to bother him as little as possible with any of this. Ideally, the current Rule UI should be sufficient for him, so the goal is rather to provide module types with clean interfaces, which allow auto-mapping of modules.
Manual mapping is rather a feature for developers, who create complex templates that might be using more generic and versatile module types internally.

**Comment by danchom** [comment]
>Manual mapping is rather a feature for developers, who create complex templates that might be using more generic and versatile module types internally.

I'm concerning this won't be possible, because the auto mapping will overwrite complex logic of template.

**Comment by kaikreuzer** [comment]
Who ever claimed that auto-mapping should overwrite anything?

**Comment by danchom** [comment]
In case when auto mapping is applied on every update it can overwrite such rules. Also the template rule is converted into regular rule and auto mapping can add additional connections.

**Comment by kaikreuzer** [comment]
As stated in https://github.com/eclipse/smarthome/issues/886#issuecomment-188330967:
"the user automatically has priority, as no auto-mapping will take place if he already created connections himself"

**Comment by danchom** [comment]
Sorry may be my misunderstanding. Let summarize if I understand right?
1) Auto-mapping always will be applied on add (update) methods when a rule does not have any connections. When a connection was set, the auto mapping must be skipped. 
2) Auto mapping will do connection only when the input tags are subset of the output tags.
3) when connections can't be done by auto-mapping the rule engine will return rule without connections silently
4) in case of conflict, an input can be connected to more then one output, the auto mapping must skip this input and leave it unconnected.

**Comment by kaikreuzer** [comment]
1) ~~rule does not have any connections~~ -> input does not have a connection.
3) What do you mean by "return"? Yes, upon creation or update, the rule will simply be exactly the one that was passed as a parameter. So there is no "silence" here, it is just the normal behavior.

**Comment by danchom** [comment]
> 1) rule does not have any connections -> input does not have a connection.

rule does not have any connections I mean there are not any connections between inputs and outputs of modules of the rule. If there is at least one connection IMHO the auto mapping should be skipped because the 'user' was set his decision. 
Otherwise if we try to reconnect empty inputs on each update we go back to the overwrite complex rule  logic scenario (for example set by template) . This means the behavior of the rule won't be determinated

**Comment by kaikreuzer** [comment]
No, this doesn't make sense, see my first comment in https://github.com/eclipse/smarthome/issues/886#issuecomment-188330967.
You might be simply adding another module to the end of a rule and while the existing ones are already connected, the newly added one definitely needs to be auto-mapped then.

**Comment by danchom** [comment]
Ok 
1) Auto-mapping always will be applied on add (update) methods when a input does not have a connection. When a connection was set, the auto mapping must be skipped. 

If someone want to leave an input unconnected a new tag can be put to the input and the auto mapping will not connect it.

**Comment by danchom** [comment]
Implemented by #1125
